pypesto/visualize/sampling.py

```diff
@@ -91,6 +114,565 @@ def sampling_fval_trace(
+        summary: Dict[str, PredictionResult],
+        condition_ids: Sequence[str],
+        observable_ids: Sequence[str],
+        axes: matplotlib.axes.Axes,
+        levels: Sequence[float],
+        levels_opacity: Dict[int, float],
+        labels: Dict[str, str],
+        variables_color: Sequence[RGB],
+) -> None:
+    """Plot predicted trajectories, with subplots grouped by observable.
+
+    See :py:func:`_plot_trajectories_by_condition` for parameter descriptions.
+    """
+    # Each subplot has all data for a single observable.
+    for observable_index, observable_id in enumerate(observable_ids):
+        t0 = 0
+        ax = axes.flat[observable_index]
+        ax.set_title(f'Trajectory: {labels[observable_id]}')
+        # Each subplot is divided by conditions, with vertical lines.
+        for condition_index, _condition_id in enumerate(condition_ids):
+            if condition_index != 0:
+                ax.axvline(
+                    t0,
+                    linewidth=2,
+                    color='k',
+                )
+
+            t_max = t0
+            t_median, y_median = _get_statistic_data(
+                summary,
+                MEDIAN,
+                condition_index,
+                observable_id,
+            )
+            t_median_shifted = t_median + t0
+            ax.plot(
+                t_median_shifted,
+                y_median,
+                'k-',
+            )
+            t_max = max(t_max, *t_median_shifted)
+            for level_index, level in enumerate(levels):
+                lower_label, upper_label = [
+                    get_percentile_label(percentile)
+                    for percentile in _get_level_percentiles(level)
+                ]
+                t_lower, lower_data = _get_statistic_data(
+                    summary,
+                    lower_label,
+                    condition_index,
+                    observable_id,
+                )
+                t_upper, upper_data = _get_statistic_data(
+                    summary,
+                    upper_label,
+                    condition_index,
+                    observable_id,
+                )
+                t_lower_shifted = t_lower + t0
+                t_upper_shifted = t_upper + t0
+                # Timepoints must match, or `upper_data` will be plotted at
+                # some incorrect time points.
+                assert (np.array(t_lower) == np.array(t_upper)).all()
+                ax.fill_between(
+                    t_lower_shifted,
+                    lower_data,
+                    upper_data,
+                    facecolor=rgba2rgb(
+                        variables_color[condition_index]
+                        + [levels_opacity[level_index]]
+                    ),
+                    lw=0,
+                )
+                t_max = max(t_max, *t_lower_shifted, *t_upper_shifted)
+            t0 = t_max
+
+
+def sampling_prediction_trajectories(
```

Comment:
I strongly feel like we want to split this routine... It's more than 200 lines...
We used to have lowlevel routines in visualization, which only operate on arrays... Maybe that could help. Or just outsource some functioality to smaller private functions

Reply:
Added the suggested private functions.